### ISSUE TITLE
Segmentation Fault & Core Dump on new install for Debian Stretch

### ISSUE THREAD

**Issue by fortes** [lastpass/lastpass-cli#231]
I setup a new machine with Debian stretch, and I get a seg fault when trying to login. No other useful information, unfortunately, just:

```
Segmentation fault (core dumped)
```

Are there some flags I should add in order to get more information here? Here is my system info:

```
% uname -a
Linux localhost 3.14.0 #1 SMP PREEMPT Tue Nov 15 18:25:51 PST 2016 x86_64 GNU/Linux
```

**Comment by bcopeland** [comment]
Maybe output of ldd?  Most likely this is something to do with the linked version of openssl.

**Comment by fortes** [comment]
Here's the `ldd` output:

```
localhost:~ ldd $(which lpass) 
        linux-vdso.so.1 (0x00007ffd51521000)
        libxml2.so.2 => /usr/lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f0d5d887000)
        libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f0d5d61c000)
        libcrypto.so.1.1 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1 (0x00007f0d5d18b000)
        libcurl.so.4 => /usr/lib/x86_64-linux-gnu/libcurl.so.4 (0x00007f0d5cf0f000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0d5cb71000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f0d5c96b000)
        libicui18n.so.57 => /usr/lib/x86_64-linux-gnu/libicui18n.so.57 (0x00007f0d5c4f1000)
        libicuuc.so.57 => /usr/lib/x86_64-linux-gnu/libicuuc.so.57 (0x00007f0d5c149000)
        libicudata.so.57 => /usr/lib/x86_64-linux-gnu/libicudata.so.57 (0x00007f0d5a6cc000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f0d5a4b2000)
        liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f0d5a28c000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0d59f86000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0d59d69000)
        libnghttp2.so.14 => /usr/lib/x86_64-linux-gnu/libnghttp2.so.14 (0x00007f0d59b43000)
        libidn.so.11 => /lib/x86_64-linux-gnu/libidn.so.11 (0x00007f0d5990f000)
        librtmp.so.1 => /usr/lib/x86_64-linux-gnu/librtmp.so.1 (0x00007f0d596f2000)
        libssh2.so.1 => /usr/lib/x86_64-linux-gnu/libssh2.so.1 (0x00007f0d594c6000)
        libssl.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.2 (0x00007f0d5925a000)
        libcrypto.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.2 (0x00007f0d58df6000)
        libgssapi_krb5.so.2 => /usr/lib/x86_64-linux-gnu/libgssapi_krb5.so.2 (0x00007f0d58bab000)
        libkrb5.so.3 => /usr/lib/x86_64-linux-gnu/libkrb5.so.3 (0x00007f0d588d1000)
        libk5crypto.so.3 => /usr/lib/x86_64-linux-gnu/libk5crypto.so.3 (0x00007f0d5869e000)
        libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007f0d5849a000)
        liblber-2.4.so.2 => /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2 (0x00007f0d58289000)
        libldap_r-2.4.so.2 => /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2 (0x00007f0d58038000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0d5de89000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f0d57cb7000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f0d57aa0000)
        libgnutls.so.30 => /usr/lib/x86_64-linux-gnu/libgnutls.so.30 (0x00007f0d5773b000)
        libhogweed.so.4 => /usr/lib/x86_64-linux-gnu/libhogweed.so.4 (0x00007f0d57504000)
        libnettle.so.6 => /usr/lib/x86_64-linux-gnu/libnettle.so.6 (0x00007f0d572cd000)
        libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f0d5704a000)
        libgcrypt.so.20 => /lib/x86_64-linux-gnu/libgcrypt.so.20 (0x00007f0d56d3b000)
        libkrb5support.so.0 => /usr/lib/x86_64-linux-gnu/libkrb5support.so.0 (0x00007f0d56b2f000)
        libkeyutils.so.1 => /lib/x86_64-linux-gnu/libkeyutils.so.1 (0x00007f0d5692b000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f0d56712000)
        libsasl2.so.2 => /usr/lib/x86_64-linux-gnu/libsasl2.so.2 (0x00007f0d564f7000)
        libp11-kit.so.0 => /usr/lib/x86_64-linux-gnu/libp11-kit.so.0 (0x00007f0d56292000)
        libtasn1.so.6 => /usr/lib/x86_64-linux-gnu/libtasn1.so.6 (0x00007f0d5607f000)
        libgpg-error.so.0 => /lib/x86_64-linux-gnu/libgpg-error.so.0 (0x00007f0d55e6b000)
        libffi.so.6 => /usr/lib/x86_64-linux-gnu/libffi.so.6 (0x00007f0d55c60000)
```

**Comment by bcopeland** [comment]
On Tue, Dec 20, 2016 at 05:32:25AM -0800, Filipe Fortes wrote:
> Here's the `ldd` output:
> 
> ```
> localhost:~ ldd $(which lpass) 
>         linux-vdso.so.1 (0x00007ffd51521000)
>         libxml2.so.2 => /usr/lib/x86_64-linux-gnu/libxml2.so.2 (0x00007f0d5d887000)
>         libssl.so.1.1 => /usr/lib/x86_64-linux-gnu/libssl.so.1.1 (0x00007f0d5d61c000)

^^ so libssl.so.1.1 is linked

>         libssl.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libssl.so.1.0.2 (0x00007f0d5925a000)
>         libcrypto.so.1.0.2 => /usr/lib/x86_64-linux-gnu/libcrypto.so.1.0.2 (0x00007f0d58df6000)

^^^ and so is 1.0.2.

This will cause no end of issues.

One of these is likely linked to libcurl-openssl.

-- 
Bob Copeland %% http://bobcopeland.com/

**Comment by fortes** [comment]
I'm a little out of my element here, but poking around it seems Debian stretch has a bunch of packages that bring in 1.1 while others bring in 1.0.

`python-cryptography` (used by a bunch of things) requires `libssl1.1`, `libcurl3` depends on `libssl1.0.2` -- I'm sure there are others, but that was just a quick glance. Removing either means removing a lot of core dependencies -- removing 1.0 removes `ssh` and 1.1 removes `openssl` :/

Is there any way to force `lpass` to use only one version?

**Comment by bcopeland** [comment]
There are a couple of different approaches here:
* you can edit CMakeLists.txt to say `find_package(OpenSSL 1.0 REQUIRED)` (I didn't do this in the repository because in principle we should work with libssl-1.1) and rebuild
* you can make sure that pkg-config reports only the appropriate version while leaving the library files intact (e.g., you can uninstall the libssl-1.1-dev package and keep the libssl-1.0-dev package). This means newly compiled things will likely use the old version.
* you can tell cmake which one to use, by either going into `ccmake .` and manually setting paths, or from the command line by doing `cmake -D OPENSSL_ROOT_DIR=/path/to/desired/openssl`
* you could rebuild libcurl with the newer ssl version, although I guess that would probably be the most work since a number of things would depend on it

**Comment by fortes** [comment]
I tried the first option (editing `CMakeLists.txt` but I see the following warning when building:

```
/usr/bin/ld: warning: libssl.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libssl.so.1.1
/usr/bin/ld: warning: libcrypto.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libcrypto.so.1.1
```

Here's the diff to `CMakeLists.txt`, in case I messed that up:

```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -30,7 +30,7 @@ endif()
 find_package(LibXml2 REQUIRED)
 include_directories(${LIBXML2_INCLUDE_DIR})

-find_package(OpenSSL REQUIRED)
+find_package(OpenSSL 1.0 REQUIRED)
 include_directories(${OPENSSL_INCLUDE_DIR})

 find_package(CURL REQUIRED)
```

**Comment by bcopeland** [comment]
Can you also remove `CMakeCache.txt`?

**Comment by fortes** [comment]
I removed `CMakeCache.txt` and the `CMakeFiles` directory, still get this error:

```
/usr/bin/ld: warning: libssl.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libssl.so.1.1
/usr/bin/ld: warning: libcrypto.so.1.0.2, needed by /usr/lib/gcc/x86_64-linux-gnu/6/../../../x86_64-linux-gnu/libcurl.so, may conflict with libcrypto.so.1.1
```

**Comment by bcopeland** [comment]
OK, perhaps cmake is still choosing 1.1 since it's >= 1.0.   Oh, did you also run `make clean` after regenerating the makefile?  I don't have a setup where I can really test this, so if that didn't work can you try one of the other options?

**Comment by fortes** [comment]
I did run `make clean`, no avail.

> you can make sure that pkg-config reports only the appropriate version while leaving the library files intact (e.g., you can uninstall the libssl-1.1-dev package and keep the libssl-1.0-dev package). This means newly compiled things will likely use the old version.

How do I change `pkg-config`? Here is current output:

```
localhost:~/.local/source/lastpass-cli master % pkg-config libssl --modversion
1.1.0c
```

**Comment by fortes** [comment]
I setup a new machine, and it now compiles and links correctly. No idea what changed, but closing.

**Comment by bcopeland** [comment]
Ok, thanks for the update -- I was actually just installing stretch to try to reproduce, but I guess I'll skip that now :)

**Comment by fortes** [comment]
Yeah, no idea what happened. Hopefully it was some issue fixed in stretch and doesn't pop up again

**Comment by gricey** [comment]
FWIW, you need to install the libssl1.0-dev package before trying to compile (with your changes to CMakeLists.txt made). That way it will find and select the lower library. Without that, you still get a successful compile, but it is built against the newest version of SSH (which still meets the criteria of 1.0 as it's higher) and will segfault when trying to log in.

**Comment by cmenning** [comment]
I just spent two hours troubleshooting this. I hope this helps anyone else who stumbles across it. In my case, it was an issue with the libssl-dev package.

I am running Mint 18.3, based on Ubuntu 16.04 (so I imagine the same problem applies to 16.04). 

In the base repos, libssl-dev is for version 1.0.2. However, in order to install newer versions of PHP I added the `ondrej/php` PPA which updates libssl-dev to version 1.1.0. This led to a conflict where libcurl is compiled against 1.0.0 and lastpass-cli gets compiled against 1.1.0. 

I fixed it by running the following:

- `apt-cache policy libssl-dev` - this shows me the installation candidates
- `sudo apt-get install libssl-dev=1.0.2g-1ubuntu4.12` - downgrades the installed version
- `make clean && make && sudo make install`
- `sudo apt-get upgrade` - returns to the newer version of libssl-dev if you want to
